From d7b7a1f12a96d7b058657493365e2b120b5995a1 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:41 -0400
Subject: [fs] nfsd: nfsd4_locku() must reference the lock stateid

Message-id: <1410898849-22126-98-git-send-email-bfields@redhat.com>
Patchwork-id: 95836
O-Subject: [Patch RHEL7 097/165] nfsd: nfsd4_locku() must reference the lock stateid
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Ensure that nfsd4_locku() keeps a reference to the lock stateid
until it is done working with it. Necessary step toward client_mutex
removal.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 858cc57336dd98ca54dff417b55a86aa101f5fb0
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 1ddf4da..4f19145 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -5147,10 +5147,12 @@ nfsd4_locku(struct svc_rqst *rqstp, struct nfsd4_compound_state *cstate,
      &stp, nn);
  if (status)
   goto out;
+ /* FIXME: move into nfs4_preprocess_seqid_op */
+ atomic_inc(&stp->st_stid.sc_count);
  filp = find_any_file(stp->st_stid.sc_file);
  if (!filp) {
   status = nfserr_lock_range;
-  goto out;
+  goto put_stateid;
  }
  file_lock = locks_alloc_lock();
  if (!file_lock) {
@@ -5180,6 +5182,8 @@ nfsd4_locku(struct svc_rqst *rqstp, struct nfsd4_compound_state *cstate,
  memcpy(&locku->lu_stateid, &stp->st_stid.sc_stateid, sizeof(stateid_t));
 fput:
  fput(filp);
+put_stateid:
+ nfs4_put_stid(&stp->st_stid);
 out:
  nfsd4_bump_seqid(cstate, status);
  if (!cstate->replay_owner)
-- 
1.7.1